### ISSUE TITLE
categorical data

### ISSUE THREAD

**Issue by geoHeil** [Microsoft/LightGBM#46]
Hi,

so far dummy coding categorical data worked fine. However, with the latest update, I see
`Feature 3050 only contains one value, will be ignored` 
in the logs for each encoded column.

This only seems to be a problem when `is_unbalance=true` is passed
How should I handle categorical data for lightGBM so that it works fine?

**Comment by guolinke** [comment]
Hi, @geoHeil , you had coverted categorical data to flatten 0/1 features, right?
And I want to know which repo version is fine?
Thanks

**Comment by geoHeil** [comment]
Current repo is fine if the parameter is set to false.

It was the first time I tried to set it to true.

Indeed 01 dummy coding was used.
Guolin Ke <notifications@github.com> schrieb am Fr. 28. Okt. 2016 um 18:12:

**Comment by guolinke** [comment]
@geoHeil Got it, i will figure out why. Thanks

**Comment by guolinke** [comment]
@geoHeil can you send me a sample of data that can re-product this ? Thanks, you can send it to my email.

**Comment by guolinke** [comment]
"is_unbalance=true" don't have any effect for feature extraction. So I have no idea what cause this.
can you provide more details to me ? like the parameters you set, and a sample of data that can reproduce this

**Comment by geoHeil** [comment]
I can hand you the paramters:

```
GBMClassifier(exec_path="/path/toLightGBM", 
                              num_iterations=4000, learning_rate=0.01,
                              num_leaves=2000, min_data_in_leaf=100, metric='binary_logloss',
                              feature_fraction=0.7, bagging_fraction=0.7, bagging_freq=0,
                              metric_freq=1, early_stopping_round=20,
                              tree_learner="feature", num_threads=4,
                              is_unbalance=True)
```

**Comment by guolinke** [comment]
If you cannot provide data.. can you provide more details about it? 
like #data, #num_category_of_that_feature, the distribution of these categories.. 
Or can you generate a fake data that re-produce this? 
One other suggestion is change "50000" to a bigger number in following lines:
https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset.cpp#L219
https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset.cpp#L231
Then compile and try again.

**Comment by geoHeil** [comment]
Please find a jupyter notebook here https://github.com/geoHeil/pythonQuestions/blob/master/lightGBM.ipynb 

Indeed, it will probably not be possible to share the data.
Can you help me improve this minimal example?

**Comment by geoHeil** [comment]
This binary https://www.dropbox.com/s/k0wmfdk747km1pr/lightgbm?dl=0 compiled around lunch the day before yesterday works **fine** - even with unbalanced = True I think this is the hash 2af0dccd692d85fce1491c484c644bbf44d36f19

For me 3a06ce35ed3dfd882b246880f32866337f31ac99 seems to break something.

From that on I only receive

```
[LightGBM] [Fatal] [LightGBM] [Fatal] [LightGBM] [Fatal] [LightGBM] [Fatal] [LightGBM] [Fatal] [LightGBM] [Fatal] input format error, should be LibSVM[LightGBM] [Fatal] [LightGBM] [Fatal] input format error, should be LibSVMinput format error, should be LibSVMinput format error, should be LibSVMinput format error, should be LibSVMinput format error, should be LibSVM
input format error, should be LibSVMinput format error, should be LibSVM
```

**Comment by geoHeil** [comment]
@guolinke I found a reproducible example: https://github.com/ArdalanM/pyLightGBM/blob/master/notebooks/regression_example_kaggle_allstate.ipynb

errors with the LIBSVM error from above for the latest build -- with the old one it works fine.

**Comment by guolinke** [comment]
@geoHeil I have try the https://github.com/ArdalanM/pyLightGBM/blob/master/notebooks/regression_example_kaggle_allstate.ipynb and it works fine...
can you try to delete your local directory of LightGBM, and re-clone the latest one, then build a new binary? 
BTW, what is the OS you run with? windows or Ubuntu?

**Comment by geoHeil** [comment]
I can try that. Running on osx

**Comment by guolinke** [comment]
@geoHeil I fixed the bug for the prediction of libsvm. can you try again?
BTW, can you work with categorical feature now?

**Comment by geoHeil** [comment]
As mentioned above: the SVM problem is fixed.
Regarding the categorical data: 

```
GBMClassifier(exec_path="path/lightgbm", 
                              num_iterations=60, learning_rate=0.1,
                              num_leaves=2000, min_data_in_leaf=10, metric='binary_logloss',
                              feature_fraction=0.7, bagging_fraction=0.7, bagging_freq=0,
                              metric_freq=1, early_stopping_round=20,
                              tree_learner="serial", is_unbalance=True, 
                              num_threads=4)
```
fails to predict a positive class if is_unbalance=True, however "works" / at least predicts more than 0 postivive class labels if this option is set to false. This seems to be strange for me.

**Comment by guolinke** [comment]
@geoHeil 
1. can you give me some comparison for AUC/Logloss for is_unbalance=True and is_unbalance=False?
2. Is all Column ignored?

**Comment by geoHeil** [comment]
Regarding 1) I will need to rund some more comparison computations over the next couple of days.
Regarding 2) only some columns are ignored.

**Comment by geoHeil** [comment]
What I can say so far: the serial learner results of no positive class labels being predicted (probably due to the warnings / errors above). In fact, the serial learner seems to stop after these errors and only the feature parallel learner continues below:
```
Feature Column_3130 only contains one value, will be ignored
[LightGBM] [Info] Finish loading data, use 5.299428 seconds
(Positive labels and negative labels are equalized via SMOTE)
[LightGBM] [Info] Number of postive:120298,  number of negative:120298
[LightGBM] [Info] Number of data:240596, Number of features:2566
[LightGBM] [Info] Finish training initilization.
[LightGBM] [Info] Start train ...
[LightGBM] [Info] cannot find more split with gain = 0.000000 , current #leaves=921
```
The feature parallel learner on a single machine predicts more than 0 positive labels. 

Shouldn't the outcome be more similar?
Anyway: regarding `[LightGBM] [Error] Feature Column_21 only contains one value, will be ignored` does lightgbm use 0 or 1 indexed data? would there be a way of obtaining the right label for such a column?

**Comment by guolinke** [comment]
1. you are using single machine, right? the result on serial_tree_learner should be same with feature_tree_learner. Are the serial_tree_learner really stop?
2. some of columns are ignored have not much effect for the training. If one column is ignore, it means this column is very sparse. if you still want to use them, you can edit "50000" in https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset.cpp#L219 to a bigger value(e.g. 500000). I will change this to the configurable parameter.
3. can you update to the latest code? i fix one bug in un_balance training.

**Comment by geoHeil** [comment]
I will do the update. Do you believe that is_unbalance=True is better than an apriori SMOTE?
Anyway: serial and feature seem to generate different results on my single machine.

**Comment by chivee** [comment]
hi @geoHeil , it's very difficult for us to reproduce this problem. could you shared some sampled data to us, especially `Feature Column_21 `.

**Comment by geoHeil** [comment]
Please have a careful look at https://github.com/geoHeil/pythonQuestions/blob/master/lightGBM.ipynb I found a reproducible example

**Comment by chivee** [comment]
@geoHeil , the linker wasn't reachable.

**Comment by geoHeil** [comment]
@chivee , but you can reach the link now?

**Comment by chivee** [comment]
@geoHeil , Still not, 404 error

**Comment by geoHeil** [comment]
@chivee sorry - but it should work now. If the problem persists I will e-mail you a jupyter notebook.

**Comment by guolinke** [comment]
@geoHeil , I still not sure about your problems. Try to summarize in following:
1. training result by "serial_tree_learner" and "feature_tree_learner" are different ?
2. set "is_unbalance=true" cannot predict positive class? 
3. Some features are ignored? (actually, this is not a bug)
Am I right?

**Comment by geoHeil** [comment]
@guolinke these are my observations. But regarding 3) this is a problem for me as I have a lot of dummy coded variables which are binary encoded and thus "ignored".

**Comment by guolinke** [comment]
@geoHeil .
The solution for 3 is  edit "50000" in https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset.cpp#L219 to a bigger value(e.g. 500000). Had you tried this?

**Comment by geoHeil** [comment]
I need to check this change.
But the column definitely contain non zero values as shown in the last
cells of the jupyter notebook. However it may be just a single value with 1
all others of zeros.
Guolin Ke <notifications@github.com> schrieb am Mi. 2. Nov. 2016 um 01:23:

**Comment by chivee** [comment]
@geoHeil ,` [LightGBM] [Error] Feature Column_99 only contains one value, will be ignored` are expected to be trigger when all samples have the same value on one features ( in this case Column_99),  so that LightGBM would not use this feature for training.
(may be we should change this log to Warning so that it will not be confusing anymore)

`is_unbalanced=true` should not used for your case for that your training data have equal amount of positive sample and negative samples. But it should  have almost same result when enabled in your case.

I have checked your jupyter notebook, which use total of 80,000 sample to training and predict on 20,000. and the number of positive sample are 10592. the result are fine to me considering the distribution over your data.

**Comment by geoHeil** [comment]
I checked again and found the issue.
As a `pd.getDummies(X)` worked e.g. without the error messages of lightGBM I continued to search for the problem and found it in my pipeline. The strange thing was that even here only 0-s and 1-s were in the columns.

The problem was that the effect coded categories were included in the pipeline as well.

Indeed here in this notebook, it was fine = always some positive labels are predicted. But I had  problems with my real data.
